server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -738,14 +751,14 @@ public boolean isReadOnly() {
         return readOnly;
     }
 
-    protected void writeIndexGen(final RepositoryData repositoryData, final long repositoryStateId) throws IOException {
+    protected void writeIndexGen(final RepositoryData repositoryData, final long expectedGen) throws IOException {
         assert isReadOnly() == false; // can not write to a read only repository
-        final long currentGen = latestIndexBlobId();
-        if (currentGen != repositoryStateId) {
+        final long currentGen = repositoryData.getGenId();
+        if (currentGen != expectedGen) {
```

Comment:
Ok, so the behavior of re-reading the last index-N file before writing a new one is kept but it is now the responsibility of the caller of `writeIndexGen()` to pass an up to date `RepositoryData`. While I'm OK with the change I find it more difficult now to identify the "generation" of the RepositoryData to be written. 

Maybe we could change `RepositoryData` so that it increments its own generation when a snapshot is removed or added and here we check that repositoryData.getGenId() == (expectedGen + 1)?

Reply:
@tlrx yea, I think that would be a nice refactoring. Unfortunately, when I last tried doing that (the auto-increment) it ran into a huge number of failing tests. I think it's worthwhile, especially with this change, but probably best left to another PR?

Reply:
> I think it's worthwhile, especially with this change, but probably best left to another PR?

I'm OK if it's done in a follow up PR